### ISSUE TITLE
cluster_kmeans fails with n_jobs=-1

### ISSUE THREAD

**Issue by cgseitz** [markovmodel/PyEMMA#1486]
Hello,

I am having a simple but strange issue. When I run the cluster_kmeans function with n_jobs=-1, it finishes quickly and produces an output, but this output cannot be read by dtrajs:
```
clk_2009g = coor.cluster_kmeans(tica_output_2009g, k=300, max_iter=70, n_jobs=-1, fixed_seed=True)
dtrajs_2009g = clk_2009g.dtrajs
```
```
TypeError                                 Traceback (most recent call last)
<ipython-input-130-e21a884f8c93> in <module>
     11 #2009 glycosylated
     12 clk_2009g = coor.cluster_kmeans(tica_output_2009g, k=300, max_iter=70, n_jobs=-1, fixed_seed=True)
---> 13 dtrajs_2009g = clk_2009g.dtrajs
     14 pplt.plot_free_energy(total_simulation_2009g[:, 0], total_simulation_2009g[:, 1], nbins=50)
     15 plot(clk_2009g.clustercenters[:,0], clk_2009g.clustercenters[:,1], linewidth=0, marker='o', color='black')

~/miniconda3/envs/pyemma/lib/python3.9/site-packages/pyemma/coordinates/clustering/interface.py in dtrajs(self)
     98         """Discrete trajectories (assigned data to cluster centers)."""
     99         if len(self._dtrajs) == 0:  # nothing assigned yet, doing that now
--> 100             self._dtrajs = self.assign(stride=1)
    101 
    102         return self._dtrajs  # returning what we have saved

~/miniconda3/envs/pyemma/lib/python3.9/site-packages/pyemma/coordinates/clustering/interface.py in assign(self, X, stride)
    214             skip = self.skip if hasattr(self, 'skip') else 0
    215             # map to column vectors
--> 216             mapped = self.get_output(stride=stride, chunk=self.chunksize, skip=skip)
    217             # flatten and save
    218             self._dtrajs = [np.transpose(m)[0] for m in mapped]

~/miniconda3/envs/pyemma/lib/python3.9/site-packages/pyemma/coordinates/data/_base/transformer.py in get_output(self, dimensions, stride, skip, chunk)
    224             self.estimate(self.data_producer, stride=stride)
    225 
--> 226         return super(StreamingTransformer, self).get_output(dimensions, stride, skip, chunk)
    227 
    228 

~/miniconda3/envs/pyemma/lib/python3.9/site-packages/pyemma/coordinates/data/_base/datasource.py in get_output(self, dimensions, stride, skip, chunk)
    404             pg.register(it.n_chunks, description='getting output of %s' % self.__class__.__name__)
    405             with pg.context(), it:
--> 406                 for itraj, chunk in it:
    407                     i = slice(it.pos, it.pos + len(chunk))
    408                     assert i.stop - i.start > 0

~/miniconda3/envs/pyemma/lib/python3.9/site-packages/pyemma/coordinates/data/_base/datasource.py in __next__(self)
   1047         self._select_file(self._itraj)
   1048         try:
-> 1049             X = self._use_cols(self._next_chunk())
   1050             self._t += len(X)
   1051         except StopIteration as e:

~/miniconda3/envs/pyemma/lib/python3.9/site-packages/pyemma/coordinates/data/_base/datasource.py in _next_chunk(self)
   1169         # We discard the trajectory index here for transformation
   1170         if self.transform_function is not None:
-> 1171             x = self.transform_function(x)
   1172         return x
   1173 

~/miniconda3/envs/pyemma/lib/python3.9/site-packages/pyemma/coordinates/clustering/interface.py in _transform_array(self, X)
    165             self._precentered = True
    166 
--> 167         dtraj = self._inst.assign(X, self.clustercenters, self.n_jobs)
    168         res = dtraj[:, None]  # always return a column vector in this function
    169         return res

TypeError: assign(): incompatible function arguments. The following argument types are supported:
    1. (self: pyemma.coordinates.clustering._ext.ClusteringBase_f, arg0: numpy.ndarray[float32], arg1: numpy.ndarray[float32], arg2: int) -> numpy.ndarray[int32]

Invoked with: <pyemma.coordinates.clustering._ext.kmeans.Kmeans_f object at 0x7fe563082870>, array([[ -2.544097  , -11.808595  ,  15.822456  , ...,  24.746307  ,
         -2.0871782 ,  -3.536114  ],
       [ -2.2522852 , -12.71603   ,  15.064904  , ...,  23.657152  ,
         -1.0526323 ,  -1.4235435 ],
       [ -2.2182655 , -12.331137  ,  15.543913  , ...,  22.21831   ,
         -1.0758259 ,  -2.0820112 ],
       ...,
       [ -0.94185764, -11.698585  ,  16.597183  , ...,  21.410278  ,
         -0.30680218,  -0.47341728],
       [ -1.5341697 , -11.96187   ,  16.401154  , ...,  21.408659  ,
         -1.9111779 ,  -1.0503037 ],
       [ -1.3355348 , -11.703044  ,  16.651745  , ...,  19.865585  ,
         -0.93530536,  -0.38585198]], dtype=float32), array([[ -2.572739  , -13.60588   ,  15.819757  , ...,  20.398506  ,
          1.3291725 ,  -0.67584485],
       [ -8.541097  , -17.111874  ,  -7.0049343 , ...,  58.456314  ,
        -26.399445  , -10.130635  ],
       [  0.62840337,  -4.2190905 ,  10.408807  , ...,   6.30783   ,
         -1.0868006 ,  -0.09319047],
       ...,
       [ -2.5245473 , -11.49325   ,  15.615449  , ...,  23.621553  ,
         -5.5561647 ,  -4.8761907 ],
       [ -2.7140334 , -13.124461  ,  15.664335  , ...,  21.7391    ,
         -1.5655806 ,  -2.872823  ],
       [ -2.0110402 , -12.242092  ,  17.700834  , ...,  21.99463   ,
          0.16626458,  -2.2592657 ]], dtype=float32), -1
```
However, when I run the exact same function setting n_jobs to a positive number, it runs far slower but successfully completes. The run speed is far slower even when I set n_jobs to be the number of processors on my machine, which I believed was the same as using n_jobs=-1. I can easily 
[pyemma_conda_list.txt](https://github.com/markovmodel/PyEMMA/files/5868941/pyemma_conda_list.txt)
run these functions setting n_jobs to some number other than -1, but the naïve user may incorrectly think they have an issue with their files instead. It is useful to note that in most to all other instances in my jupyter notebook I have used n_jobs=-1 without issue.

I have attached the conda environment but the "tica_output_2009g" data is huge (256 MB) even after zipping, so that cannot be attached. I am using pyemma version 2.5.7 and this OS info:
```
NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

**Comment by thempel** [comment]
I agree that there should be no random output generated. However, I'm not sure why `n_jobs=-1` is supposed to be a shortcut for "all resources"? Is that documented somewhere? In the [documentation](http://www.emma-project.org/latest/api/generated/pyemma.coordinates.cluster_kmeans.html#pyemma.coordinates.cluster_kmeans) of the k-means method I find this:

> n_jobs (int or None, default None) – Number of threads to use during assignment of the data. If None, all available CPUs will be used.

**Comment by clonker** [comment]
Setting `n_jobs` to `-1` is indeed not supported. In the background the parallelization is performed using OpenMP and also there setting a number of threads to `-1` is [not supported](https://www.openmp.org/spec-html/5.0/openmpsu110.html), at best implementation-defined. Setting it to `None` (which is the default) uses the number of threads equal to the number of physical cores. You may want to take the number of physical cores times 2 or times 4 to achieve some speedup because the operating system can do a better job at load balancing and make use of hyper-threading.
The error-handling could be improved though and especially the error message you have been left with is anything but descriptive. Perhaps we also should think about a more sensible default which `None` resolves to (e.g. `2*n_physical_cores`  to account for hyper threading).

**Comment by cgseitz** [comment]
Thanks for the info, Moritz and Tim. I set n_jobs=-1 because this [link](http://www.emma-project.org/latest/legacy-notebooks/applications/pentapeptide_msm/pentapeptide_msm.html?highlight=n_jobs) used it ([this legacy one does too](http://www.emma-project.org/latest/legacy-notebooks/applications/trypsin_benzamidine_hmm/trypsin_benzamidine_hmm.html?highlight=n_jobs)) and using it for other pyemma functions did not return errors so I did not notice that it wasn't supported. I didn't realize that pyemma supported hyperthreading; the software I was using for a different project did not so it had fallen off my radar. Setting that as the default or putting it in the documentation would be nice - none of the several other pyemma users in my lab were aware that pyemma can hyperthread, so I will spread the word.

**Comment by clonker** [comment]
Oh that is unfortunate. Also there it is not explicitly supported and falls back to 'no parallelization' if you set it to `-1`. So we have 

- [x] ... a bug in the documentation,
- [x] ... inconsistent behavior with respect to n_jobs between estimating implied timescales and e.g. k-means (we should probably check the other places it is used, too),
- [x] ... a better job to do at argument validation, raising an error early.

Thanks for bringing this to our attention!

**Comment by marscher** [comment]
The pattern of n_jobs has been inspired by scikit-learn (e.g. https://github.com/scikit-learn/scikit-learn/search?p=2&q=n_jobs). There -1 means using all available processors, at least in lots of examples in sklearn. I'd recommend to treat -1 and None as arguments to use all available resources. This wouldn't break older scripts passing -1, expecting that full parallelization is used.
But I also agree with Tim, that -1 is not really intuitive.

**Comment by clonker** [comment]
Scikit-learn just deprecated n_jobs altogether. I agree that for the sake of backward-compatibility, -1 could/should be treated as None, on the other hand the backport of deeptime functionality likely warrants a new major version anyways and at least in OpenMP terms, -1 is undefined behavior.

**Comment by clonker** [comment]
So I think the best compromise we can strike is using a deprecation warning instead of a hard raise, I just changed that in the PR.

**Comment by clonker** [comment]
Implemented in PR #1488
